### ISSUE TITLE
Timeout and Retry Policies with Request Bodies

### ISSUE THREAD

**Issue by jdonofrio728** [failsafe-go/failsafe-go#73]
Perhaps I'm doing something wrong, but I am unable to compose a policy with timeouts and retries that utilize a HTTP Post + Request body.

```
package main

import (
	"bytes"
	"fmt"
	"github.com/failsafe-go/failsafe-go"
	"github.com/failsafe-go/failsafe-go/failsafehttp"
	"github.com/failsafe-go/failsafe-go/timeout"
	"io"
	"net/http"
	"net/http/httptest"
	"sync/atomic"
	"time"
)

func main() {
	retryPolicy := failsafehttp.RetryPolicyBuilder().
		WithMaxRetries(3).
		OnRetry(func(e failsafe.ExecutionEvent[*http.Response]) {
			// What do we do when we retry?
			fmt.Printf("Retry ElapsedAttemptTime: %v lastError: %#v, isRetry:%v\n", e.ElapsedAttemptTime(), e.LastError(), e.IsRetry())
		}).
		OnAbort(func(e failsafe.ExecutionEvent[*http.Response]) {
			// What do we do when we abort?
			fmt.Println("Abort")
		}).
		OnSuccess(func(e failsafe.ExecutionEvent[*http.Response]) {
			// What do we do when we are successful?
			fmt.Println("Success")
		}).
		Build()

	timeoutPolicy := timeout.Builder[*http.Response](100 * time.Millisecond).
		OnTimeoutExceeded(func(e failsafe.ExecutionDoneEvent[*http.Response]) {
			// Timed out
			fmt.Println("Timeout")
		}).
		Build()

	// Start test server
	slow := slowServer(1 * time.Second)
	defer slow.Close()

	// Create the http request
	req, err := http.NewRequest(http.MethodGet, slow.URL, bytes.NewReader([]byte("blah")))
	// req, err := http.NewRequest(http.MethodGet, slow.URL, nil)

	if err != nil {
		panic(err)
	}

	rt := failsafehttp.NewRoundTripper(nil, retryPolicy, timeoutPolicy)
	client := &http.Client{
		Transport: rt,
	}

	resp, err := client.Do(req)
	if err == nil {
		resp.Body.Close()
	}
}

func slowServer(delay time.Duration) *httptest.Server {
	slowTimes := atomic.Int32{}
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, request *http.Request) {
		slowTimes.Add(1)
		fmt.Println("got request", slowTimes.Load())
		payload, err := io.ReadAll(request.Body)
		request.Body.Close()
		if err != nil {
			fmt.Println(err)
		}
		fmt.Printf("body=%s\n", string(payload))
		if slowTimes.Load() == 2 {
			// After two slow requests, go fast
			fmt.Fprintf(w, "pong")
		}
		timer := time.After(delay)
		select {
		case <-request.Context().Done():
			fmt.Println("request done")
		case <-timer:
			fmt.Fprintf(w, "pong")
		}
	}))
}
```
I think what is happening is that the request body "blah" is read on the reader once, then on subsequent attempts, the reader doesn't return anything to send and errors out while building the request.

```
# go run .
got request 1
body=blah
Timeout
request done
Retry ElapsedAttemptTime: 12.458µs lastError: &errors.errorString{s:"timeout exceeded"}, isRetry:true
Retry ElapsedAttemptTime: 875ns lastError: http.nothingWrittenError{error:(*errors.errorString)(0x140001801e0)}, isRetry:true
Retry ElapsedAttemptTime: 667ns lastError: http.nothingWrittenError{error:(*errors.errorString)(0x14000180320)}, isRetry:true
```

**Comment by jhalterman** [comment]
Thanks for filing this @jdonofrio728. I just pushed a fix to main via fc34a7f. Can you confirm this resolves things for you?

**Comment by jhalterman** [comment]
This has been fixed in 0.6.9.

**Comment by jdonofrio728** [comment]
This fixed the test I provided. Thanks!
